docs/user/tutorial-asgi.rst

```diff
@@ -0,0 +1,1090 @@
+                resp.set_header(self.CACHE_HEADER, 'Hit')
+            else:
+                resp.set_header(self.CACHE_HEADER, 'Miss')
+
+        async def process_response(self, req, resp, resource, req_succeeded):
+            if not req_succeeded:
+                return
+
+            key = f'{self.PREFIX}/{req.path}'
+
+            if req.method in self.INVALIDATE_ON:
+                await self.redis.delete(key)
+            elif not resp.context.cached:
+                data = await self.serialize_response(resp)
+                await self.redis.set(key, data, expire=self.TTL)
+
+For caching to come into effect, we also need to add the ``RedisCache``
+component to our application's middleware list.
+The final definition of all components in ``app.py`` now is:
+
+.. Copy-paste under: examples/asgilook/asgilook/app.py
+
+.. code:: python
+
+    import falcon.asgi
+
+    from .cache import RedisCache
+    from .config import Config
+    from .images import Images, Thumbnails
+    from .store import Store
+
+
+    def create_app(config=None):
+        config = config or Config()
+        cache = RedisCache(config)
+        store = Store(config)
+        images = Images(config, store)
+        thumbnails = Thumbnails(store)
+
+        app = falcon.asgi.App(middleware=[cache])
+        app.add_route('/images', images)
+        app.add_route('/images/{image_id:uuid}.jpeg', images, suffix='image')
+        app.add_route('/thumbnails/{image_id:uuid}/{width:int}x{height:int}.jpeg',
+                      thumbnails)
+
+        return app
+
+Now, subsequent access to ``/thumbnails`` should be cached, as indicated by the
+``x-asgilook-cache`` header::
+
+  $ http localhost:8000/thumbnails/167308e4-e444-4ad9-88b2-c8751a4e37d4/115x115.jpeg
+
+  HTTP/1.1 200 OK
+  content-length: 2985
+  content-type: image/jpeg
+  date: Tue, 24 Dec 2019 19:46:51 GMT
+  server: uvicorn
+  x-asgilook-cache: Hit
+
+  +-----------------------------------------+
+  | NOTE: binary data not shown in terminal |
+  +-----------------------------------------+
+
+.. note::
+   Left as another exercise for the reader: individual images are streamed
+   directly from ``aiofiles`` instances, and caching therefore does not work
+   for them at the moment.
+
+The project's structure should now look like this::
+
+  asgilook
+  ├── .venv
+  └── asgilook
+      ├── __init__.py
+      ├── app.py
+      ├── asgi.py
+      ├── cache.py
+      ├── config.py
+      ├── images.py
+      └── store.py
+
+Testing Our Application
+-----------------------
+
+So far, so good? We have only tested our application by sending a handful of
+requests manually. Have we tested all code paths? Have we covered typical user
+inputs to the application?
+
+Having a comprehensive test suite is vital not only for verifying that
+application is correctly behaving at the moment, but also limiting the impact
+of future regressions that will have been introduced into the codebase.
```

Comment:
`will have been introduced` -> `may be introduced`

Reply:
Hm, I have actually chosen this form intentionally. Maybe I could clarify that intent?
> will have inevitably been introduced

Reply:
Oh ok. Fine either way then

Reply:
Reworded; it should hopefully read a bit easier now.